crates/iota-transaction-builder/src/builder.rs

```diff
@@ -0,0 +1,1066 @@
+            objects: vec![coin_arg],
+            address: rec_arg,
+        }));
+        self
+    }
+
+    /// Get the current set gas coins.
+    pub fn get_gas(&self) -> impl Iterator<Item = ObjectId> + '_ {
+        self.inputs.values().filter_map(|i| {
+            if i.is_gas {
+                i.object_id().copied()
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Set the gas budget. Optional.
+    pub fn gas_budget(&mut self, gas_budget: u64) -> &mut Self {
+        self.gas_budget = Some(gas_budget);
+        self
+    }
+
+    /// Set the gas price. Optional.
+    pub fn gas_price(&mut self, gas_price: u64) -> &mut Self {
+        self.gas_price = Some(gas_price);
+        self
+    }
+
+    /// Set the sponsor. Optional.
+    pub fn sponsor(&mut self, sponsor: Address) -> &mut Self {
+        self.sponsor = Some(sponsor);
+        self
+    }
+
+    /// Set the expiration. Optional.
+    pub fn expiration(&mut self, expiration: u64) -> &mut Self {
+        self.expiration = TransactionExpiration::Epoch(expiration);
+        self
+    }
+
+    /// Make a value available to the transaction as an input.
+    fn input(&mut self, kind: InputKind, is_gas: bool) -> Argument {
+        if let Some((i, input)) = self.inputs.iter_mut().find(|(_, input)| input.kind == kind) {
+            if is_gas {
+                input.is_gas = true;
+            }
+            return Argument::Input(*i as _);
+        }
+        let idx = self
+            .inputs
+            .last_entry()
+            .map(|e| *e.key() + 1)
+            .unwrap_or_default();
+        self.inputs.insert(idx, Input { kind, is_gas });
+        Argument::Input(idx as _)
+    }
+
+    /// Add a pure input using the BCS serialized bytes
+    pub fn pure_bytes(&mut self, bytes: Vec<u8>) -> Argument {
+        self.input(
+            InputKind::Input(iota_types::Input::Pure { value: bytes }),
+            false,
+        )
+    }
+
+    /// Add a pure input
+    pub fn pure<T: Serialize>(&mut self, value: T) -> Argument {
+        // This serialization should never fail, so we will forego error propagation
+        // here for convenience
+        self.pure_bytes(bcs::to_bytes(&value).expect("bcs serialization failed"))
+    }
+
+    /// Add a new command to the PTB
+    pub fn command(&mut self, command: Command) -> Argument {
+        let i = self.commands.len();
+        self.commands.push(command);
+        Argument::Result(i as u16)
+    }
+
+    /// Manually set a command with an optional name
+    pub fn named_command(&mut self, cmd: Command, name: impl NamedCommands) {
+        self.command(cmd);
+        name.push_named_commands(self);
+    }
+
+    /// Get the value for the given string in the named commands map
+    pub fn get_named_command(&self, name: &str) -> Option<Argument> {
+        self.named_commands.get(name).copied()
+    }
+}
+
+impl TransactionBuilder<()> {
```

Comment:
Why have a separated block like this with the `new()` and `impl<C> TransactionBuilder<C> {` in between them?